### PR TITLE
Pod Wars Team Planning Maps

### PR DESCRIPTION
**Pull request by Mister-Moriarty**
[Feature] [Gamemodes] [Add To Wiki]

## About the PR:
Expands upon the physical station minimaps feature implemented in #12463 and previous PRs to work with maps of any size, and adds map displays and controller computers to the Bridge of both the NSV Pytheas and Lodbrok.

Each debris field map computer will display the location of both pod carriers, the three control points, all minor areas, asteroids, and warp beacons, and the location of every living member of the respective team. Should one of the teams capture a control point, the colour of the control point on the minimap will change to the primary colour of the capturing team. Their controllers operate in the same way as the handheld controller on the Cairngorm.

image

To this end, a minimap renderer datum has been created, which will handle all operations related to drawing the minimaps, and allow for specific areas to be drawn separately on an `atom/movable` for easy recolouring across various minimap instances. Each pod pilot will also receive a special tracking implant, and the Bridges of the two pod carriers have been redesigned.

## Why's this needed?
While static debris field maps already exist, an additional live, high visibility minimap will help aid coordination within teams, especially in matters of determining where to stage an assualt, where all players on a team are currently at, and in navigating the field.

## Changelog:
```changelog
(u)Mr. Moriarty
(*)Pod Wars: The NSV Pytheas and Lodbrok have both been installed with debris field map displays and map controllers, similar to the ones that may be found on the Cairngorm, situated in their Bridges.
```

### CODE REVIEW

File: code/modules/minimap/minimap_renderer.dm [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+			minimap_render.Blend(minimap_type_render, ICON_OVERLAY)
+
+		minimap_render_object.icon = minimap_render
+		minimap_render_object.vis_flags = VIS_INHERIT_ID | VIS_INHERIT_LAYER
+		minimap_render_object.mouse_opacity = 0
+
+		return minimap_render_object
+
+	///Recolours a specified area group to a specified colour.
+	proc/recolour_area(var/area_group, var/colour)
```
Comment (TobleroneSwordfish) [inline code/modules/minimap/minimap_renderer.dm:147]: as much as it pains me as a fellow brit, we should probably stick to `color` for proc names at least (don't worry about comments tho)

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
Extremely cool, I wasn't aware this had actual pilot tracking too. The code overall looks good, I'm a little concerned about the memory impacts of shoving every solid turf in a 500x500 z level in a list, but it seems to load fine and is only a roundstart issue anyway.

The minimap markers on the map controllers are ludicrously small currently, to the point of being 1-2 pixels (probably caused by the larger map size)
image

**Comment by Mister-Moriarty** [comment]
> ...I'm a little concerned about the memory impacts of shoving every solid turf in a 500x500 z level in a list, but it seems to load fine and is only a roundstart issue anyway.

A turf will only be placed into that list should the area containing it have a non-null `minimaps_to_render_on` var, so only a small fraction of the (likely more than) 500^2 turfs will be placed into that list.
​

> The minimap markers on the map controllers are ludicrously small currently, to the point of being 1-2 pixels (probably caused by the larger map size)

Regarding the miniscule pilot markers, that will likely require a resizing system for minimap markers, which I will hopefully begin work on soon.

Thank you for the review!

File: code/modules/minimap/minimap_object.dm [hunk 2]
```diff
@@ -342,3 +375,61 @@
 		if (minimap)
 			src.minimap_controller = new(minimap)
 			src.minimap_ui = new(src, "nukeop_map", src.minimap_controller, "Atrium Station Map Controller", "syndicate")
+
+/obj/machinery/computer/pod_wars_minimap_controller
```
Comment (ZeWaka) [inline code/modules/minimap/minimap_object.dm:379]: this should be an abstract type right?

File: code/modules/minimap/minimap_renderer.dm [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+				minimap_type_render_list["[area_group]"] = area_render_object
+
+	///Separates a given bitflag into a list of the separate bits composing it. Eg. 1011 -> (0001, 0010, 1000), or in base 10: 11 -> (1, 2, 8)
+	proc/separate_bitflag(var/bitflag)
+		var/list/flag_list = list()
+		var/bit_length = ceil(log(2, bitflag + 1))
+
+		for (var/i = 0 to bit_length)
+			if (2**i & bitflag)
```
Comment (ZeWaka) [inline code/modules/minimap/minimap_renderer.dm:97]: parentheses pls

**Comment by ZeWaka** [comment]
fuu

**Review by ZeWaka — APPROVED** [review]

**Comment by Kyle2143** [comment]
Yes, this is amazing. This is something I had "planned" to add to each ship shortly after releasing pod wars. This has literally all the features I had considered for this, so I wonder if I actually wrotw it down or this is just such a logical feature that anyone would arrive at it.

 I remember Sov pitched a good way to implement this, but I never got around to it cause I was already starting to wind down for a while and have been not here.

**Pull request merged**